lib/up4.ml

```diff
@@ -0,0 +1,251 @@
+
+  let eval_extern name =
+    match name with
+    | "set_out_port" -> eval_set_out_port
+    | "get_in_port" -> eval_get_in_port 
+    | "get_out_port" -> eval_get_out_port
+    | "get_value" -> eval_get_value
+    | "drop" -> eval_drop
+    | _ -> failwith "extern unknown in up4 or not yet implemented"
+
+  (* Use the port number [meta] to set out_port and in_port of Im_t extern object. *)
+  (* QUEUE_DEPTH_AT_DEQUEUE metadata hard coded.. it says in the up4 spec that 
+     metadata_fileds_t enums are immutable intrinsic metadata field for the target. *)
+  let initialize_metadata meta st =
+    State.insert_extern State.im_t_location (
+      Im_t {out_port = meta; in_port = meta; queue_depth_at_dequeue = Bigint.of_int 32}) st
+
+  let check_pipeline env = failwith "unimplemented"
+
+  (* below is copied from ebpf.ml *)
+  let eval_up4_ctrl (ctrl : ctrl) (control : value) (args : Expression.t option list) app
+      (env,st) : state * signal =
+    let (st,s,_) = app ctrl env st SContinue control args in 
+    (st,s)
+
+  let helper (param_string : loc) (param_type : Type.t) (env : env) (st : state)
+    : value * Types.name * loc * state * env =
+    let param_value = init_val_of_typ env param_type in 
+    let param_name = Types.BareName (Info.dummy, param_string) in
+    let param_loc = State.fresh_loc () in 
+    let st = State.insert_heap param_loc param_value st in 
+    let env = EvalEnv.insert_val param_name param_loc env in
+    let env = EvalEnv.insert_typ param_name param_type env in
+    param_value, param_name, param_loc, st, env
+
+  (* Push [pkt] through pipeline of this architecture, 
+     updating the environment and state and return the 
+     new state, environment and the packet if it was accepted! *)
+  let eval_pipeline (ctrl : ctrl) (env : env) (st : state) (pkt : pkt)
+      (app : state apply) : state * env * pkt option =
+    (* Get main *)
+    let main = State.find_heap (EvalEnv.find_val (BareName (Info.dummy, "main")) env) st in
+    (* Get arguments passed into main *)
+    let vs = assert_package main |> snd in
+    (* Get Package parameters *)
+    let parser = List.Assoc.find_exn vs "p" ~equal:String.equal |> fun x -> State.find_heap x st in
+    let control = List.Assoc.find_exn vs "c" ~equal:String.equal |> fun x -> State.find_heap x st in
+    let deparser = List.Assoc.find_exn vs "d" ~equal:String.equal |> fun x -> State.find_heap x st in
+    (* Parser params *)
+    let params =
+      match parser with
+      | VParser {pparams=ps;_} -> ps
+      | _ -> failwith "parser is not a parser object" in
+    (* Micro Control params *)
+    let control_params = 
+      match control with 
+      | VControl {cparams=ps;_} -> ps
+      | _ -> failwith "control is not a control object" in 
+    (* Deparser params *)
+    let deparse_params = 
+      match deparser with 
+      | VControl {cparams=ps;_} -> ps
+      | _ -> failwith "deparser is not a control object" in 
```

Comment:
I think there are `assert_parser`, `assert_control`, etc. functions in the Value module you could use to simplify this a bit.